scipy/stats/_hypotests.py

```diff
@@ -1262,30 +1262,46 @@ def _get_binomial_log_p_value_with_nuisance_param(
     return -log_pvalue
 
 
-def _pval_cvm_2samp_exact(s, nx, ny):
+def _pval_cvm_2samp_exact(s, m, n):
     """
     Compute the exact p-value of the Cramer-von Mises two-sample test
-    for a given value s (float) of the test statistic by enumerating
-    all possible combinations. nx and ny are the sizes of the samples.
+    for a given value s of the test statistic.
+    m and n are the sizes of the samples.
+
+    [1] Y. Xiao, A. Gordon, and A. Yakovlev, "A C++ Program for
+        the Cramér-Von Mises Two-Sample Test", J. Stat. Soft.,
+        vol. 17, no. 8, pp. 1-15, Dec. 2006.
+    [2] T. W. Anderson "On the Distribution of the Two-Sample Cramer-von Mises
+        Criterion," The Annals of Mathematical Statistics, Ann. Math. Statist.
+        33(3), 1148-1159, (September, 1962)
     """
-    rangex = np.arange(nx)
-    rangey = np.arange(ny)
 
-    us = []
-
-    # x and y are all possible partitions of ranks from 0 to nx + ny - 1
-    # into two sets of length nx and ny
-    # Here, ranks are from 0 to nx + ny - 1 instead of 1 to nx + ny, but
-    # this does not change the value of the statistic.
-    for x, y in _all_partitions(nx, ny):
-        # compute the statistic
-        u = nx * np.sum((x - rangex)**2)
-        u += ny * np.sum((y - rangey)**2)
-        us.append(u)
-
-    # compute the values of u and the frequencies
-    u, cnt = np.unique(us, return_counts=True)
-    return np.sum(cnt[u >= s]) / np.sum(cnt)
+    # [1, p. 3]
+    lcm = np.lcm(m, n)
+    # [1, p. 4], below eq. 3
+    a = lcm // m
+    b = lcm // n
+    # eq. 9 in [2] and eq. 2 in [1]
+    mn = m * n
+    zeta = lcm ** 2 * (m + n) * (6 * s - mn * (4 * mn - 1)) // (6 * mn ** 2)
```

Comment:
I am sorry for not writing any explanation about this. This can be shown using both Eq. 9 in [2] and Eq. 2 in [1].

I found `s` in the code was equal to the variable $U$ in [2] (because, `s` is the first argument of `_pval_cvm_2samp_exact`, we call `_pval_cvm_2samp_exact(u, nx, ny)` in the function `cramervonmises_2samp`, and `u` is calculated as defined in [2], Eq. 10.
We also find $T_2$ defined in [1], Eq. 1 is equivalent to $T$ defined in [2], Eq. 3.
So we have $T_2 = (mn / (m + n)^2 L^2) \zeta$ as indicated in [1] (just below Eq. 2), hence $\zeta = (L^2 (m + n)^2 / mn) T_2$. Using Eq. 9 in [2], $$\begin{align}\zeta &= \frac{L^2 (m + n)^2}{mn}\biggl(\frac{U}{mn(m + n)} - \frac{4mn - 1}{6(m + n)}\biggr) \\\\ &= \frac{L^2 (m + n)^2}{mn} \cdot \frac{6U - mn(4mn - 1)}{6mn(m + n)} \\\\ &= \frac{L^2 (m + n) (6U - mn(4 mn - 1))}{6(mn)^2}, \end{align}$$ that is the same relation as L.1286 in the code, and we can calculate this *without floating-point error*. (To calculate $T_2$, we have to use floating-point numbers.)

Reply:
Ok, that's what I thought, but my mental math was not getting me there. Thanks!